test/unittests/OpModel/TTNN/Lib/TestOpModelLib.cpp

```diff
@@ -0,0 +1,660 @@
+      tensorShape, inputLayout_l1, tensorShape, inputLayout_l1, tensorShape,
+      inputLayout_l1, false, false, workerGrid);
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 786432);
+  EXPECT_EQ(output_size, 151552);
+  EXPECT_EQ(peak_size, 151552);
+}
+
+TEST_F(OpModelBase, MatmulSharded) {
+  const llvm::SmallVector<int64_t> tensorShape = {1024, 1024};
+  const SmallVector<int64_t> gridShape = {4, 4};
+  const auto workerGrid = CreateWorkerGrid(
+      mlir::tt::ttnn::TensorMemoryLayout::BlockSharded, gridShape);
+  const mlir::tt::ttnn::TTNNLayoutAttr layout_l1_hs =
+      CreateLayout(tensorShape, mlir::tt::ttnn::BufferType::L1,
+                   mlir::tt::ttnn::TensorMemoryLayout::BlockSharded, gridShape);
+  const mlir::tt::ttnn::TTNNLayoutAttr layout_dram =
+      CreateLayout(tensorShape, mlir::tt::ttnn::BufferType::DRAM,
+                   mlir::tt::ttnn::TensorMemoryLayout::Interleaved, gridShape);
+
+  std::tuple<bool, std::optional<std::tuple<size_t, size_t, size_t>>,
+             std::optional<std::string>>
+      opConstraint;
+
+  bool legal = false;
+  std::optional<std::tuple<size_t, size_t, size_t>> l1Usage = std::nullopt;
+  std::optional<std::string> errorMsg = "";
+  size_t cb_size = 0;
+  size_t peak_size = 0;
+  size_t output_size = 0;
+
+  opConstraint = MatmulOpInterface::getOpConstraints(
+      tensorShape, layout_l1_hs, tensorShape, layout_dram, tensorShape,
+      layout_l1_hs, false, false, workerGrid);
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 524352);
+  EXPECT_EQ(output_size, 131072);
+  EXPECT_EQ(peak_size, 131072);
+
+  opConstraint = (MatmulOpInterface::getOpConstraints(
+      tensorShape, layout_l1_hs, tensorShape, layout_dram, tensorShape,
+      layout_dram, false, false, workerGrid));
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 655424);
+  EXPECT_EQ(output_size, 0); // dram
+  EXPECT_EQ(peak_size, 0);
+
+  opConstraint = (MatmulOpInterface::getOpConstraints(
+      tensorShape, layout_dram, tensorShape, layout_dram, tensorShape,
+      layout_l1_hs, false, false, workerGrid));
+  std::tie(legal, l1Usage, errorMsg) = opConstraint;
+  EXPECT_TRUE(legal);
+  EXPECT_TRUE(l1Usage.has_value());
+  EXPECT_FALSE(errorMsg.has_value());
+  std::tie(cb_size, peak_size, output_size) = l1Usage.value();
+  EXPECT_EQ(cb_size, 262144);
+  EXPECT_EQ(output_size, 524288);
+  EXPECT_EQ(peak_size, 524288);
+}
```

Comment:
Could we test a few more cases for matmul?
* Hight sharded
* Block sharded for both inputs and output?
* Mix of block / hight sharding (this shouldn't work?)